File: windows/MoDi.Desktop/Connectivity/Transports/LegacyTransportSessionAdapter.cs
using System;
using System.Collections.Generic;
using System.Threading;
using System.Threading.Tasks;
using MoDi.App.Contracts.Connectivity;
using MoDi.Core;
using MoDi.Protocol;

namespace MoDi.Desktop.Connectivity.Transports;

/// <summary>Temporary bridge from the approved legacy transport binary API to the new app-level port.</summary>
internal sealed class LegacyTransportSessionAdapter : ITransportSession
{
    private readonly ITransport _legacy;
    private readonly Func<CancellationToken, Task<IReadOnlyList<TransportCandidate>>>? _discover;
    private readonly Func<CancellationToken, Task>? _listen;
    private readonly Func<TransportEndpoint, CancellationToken, Task>? _connect;
    private readonly SemaphoreSlim _gate = new(1, 1);
    private bool _disposed;

    public LegacyTransportSessionAdapter(
        TransportDescriptor descriptor,
        ITransport legacy,
        Func<CancellationToken, Task<IReadOnlyList<TransportCandidate>>>? discover = null,
        Func<CancellationToken, Task>? listen = null,
        Func<TransportEndpoint, CancellationToken, Task>? connect = null)
    {
        Descriptor = descriptor;
        _legacy = legacy;
        _discover = discover;
        _listen = listen;
        _connect = connect;
        _legacy.PacketReceived += OnLegacyPacketReceived;
    }

    public TransportDescriptor Descriptor { get; }
    public TransportSessionState State { get; private set; } = TransportSessionState.Idle;
    public event Action<ReadOnlyMemory<byte>>? BytesReceived;

    public async Task<IReadOnlyList<TransportCandidate>> DiscoverAsync(CancellationToken cancellationToken)
    {
        ThrowIfDisposed();
        if (!Descriptor.SupportsDiscovery) return [];
        if (_discover is null) throw new InvalidOperationException($"Discovery is not bound for {Descriptor.Kind}.");
        State = TransportSessionState.Discovering;
        try { return await _discover(cancellationToken).ConfigureAwait(false); }
        finally { if (State == TransportSessionState.Discovering) State = TransportSessionState.Idle; }
    }

    public async Task ListenAsync(CancellationToken cancellationToken)
    {
        ThrowIfDisposed();
        if (!Descriptor.SupportsListening || _listen is null)
            throw new InvalidOperationException($"Listening is not bound for {Descriptor.Kind}.");
        State = TransportSessionState.Listening;
        await _listen(cancellationToken).ConfigureAwait(false);
    }

    public async Task ConnectAsync(TransportEndpoint endpoint, CancellationToken cancellationToken)
    {
        await _gate.WaitAsync(cancellationToken).ConfigureAwait(false);
        try
        {
            ThrowIfDisposed();
            if (State == TransportSessionState.Connected) return;
            State = TransportSessionState.Connecting;
            try
            {
                if (_connect is not null) await _connect(endpoint, cancellationToken).ConfigureAwait(false);
                else await _legacy.ConnectAsync(cancellationToken).ConfigureAwait(false);
                State = TransportSessionState.Connected;
            }
            catch
            {
                State = TransportSessionState.Failed;
                throw;
            }
        }
        finally { _gate.Release(); }
    }

    public Task SendAsync(ReadOnlyMemory<byte> payload, CancellationToken cancellationToken)
    {
        ThrowIfDisposed();
        if (State != TransportSessionState.Connected)
            throw new InvalidOperationException("Transport session is not connected.");
        return _legacy.SendAsync(payload, cancellationToken);
    }

    public async Task CloseAsync(CancellationToken cancellationToken)
    {
        await _gate.WaitAsync(cancellationToken).ConfigureAwait(false);
        try
        {
            if (State is TransportSessionState.Closed or TransportSessionState.Idle) { State = TransportSessionState.Closed; return; }
            State = TransportSessionState.Closing;
            await _legacy.DisconnectAsync().WaitAsync(cancellationToken).ConfigureAwait(false);
            State = TransportSessionState.Closed;
        }
        catch
        {
            State = TransportSessionState.Failed;
            throw;
        }
        finally { _gate.Release(); }
    }

    public async ValueTask DisposeAsync()
    {
        if (_disposed) return;
        await CloseAsync(CancellationToken.None).ConfigureAwait(false);
        _legacy.PacketReceived -= OnLegacyPacketReceived;
        if (_legacy is IDisposable disposable) disposable.Dispose();
        _disposed = true;
        _gate.Dispose();
    }

    private void OnLegacyPacketReceived(ReadOnlyMemory<byte> payload) => BytesReceived?.Invoke(payload);
    private void ThrowIfDisposed() => ObjectDisposedException.ThrowIf(_disposed, this);
}

File: android/app/src/main/java/com/modi/connect/core/connectivity/transport/LegacyTransportSessionAdapter.kt
package com.modi.connect.core.connectivity.transport

import com.modi.connect.core.connectivity.TransportCandidate
import com.modi.connect.core.connectivity.TransportDescriptor
import com.modi.connect.core.connectivity.TransportEndpoint
import com.modi.connect.core.connectivity.TransportSession
import com.modi.connect.core.connectivity.TransportSessionState
import com.modi.protocol.ITransport
import kotlinx.coroutines.sync.Mutex
import kotlinx.coroutines.sync.withLock

/** Temporary bridge from the approved legacy transport binary API to the new app-level port. */
class LegacyTransportSessionAdapter(
    override val descriptor: TransportDescriptor,
    private val legacy: ITransport,
    private val discoverBlock: (suspend () -> List<TransportCandidate>)? = null,
    private val listenBlock: (suspend () -> Unit)? = null,
    private val connectBlock: (suspend (TransportEndpoint) -> Unit)? = null,
) : TransportSession {
    private val gate = Mutex()
    private var disposed = false
    override var state: TransportSessionState = TransportSessionState.Idle
        private set
    override var onBytesReceived: ((ByteArray) -> Unit)? = null

    init {
        legacy.onPacketReceived = { payload -> onBytesReceived?.invoke(payload) }
    }

    override suspend fun discover(): List<TransportCandidate> {
        checkActive()
        if (!descriptor.supportsDiscovery) return emptyList()
        val operation = checkNotNull(discoverBlock) { "Discovery is not bound for ${descriptor.kind}" }
        state = TransportSessionState.Discovering
        return try { operation() } finally {
            if (state == TransportSessionState.Discovering) state = TransportSessionState.Idle
        }
    }

    override suspend fun listen() {
        checkActive()
        check(descriptor.supportsListening && listenBlock != null) { "Listening is not bound for ${descriptor.kind}" }
        state = TransportSessionState.Listening
        listenBlock.invoke()
    }

    override suspend fun connect(endpoint: TransportEndpoint) = gate.withLock {
        checkActive()
        if (state == TransportSessionState.Connected) return@withLock
        state = TransportSessionState.Connecting
        try {
            connectBlock?.invoke(endpoint) ?: legacy.connect()
            state = TransportSessionState.Connected
        } catch (error: Throwable) {
            state = TransportSessionState.Failed
            throw error
        }
    }

    override suspend fun send(payload: ByteArray) {
        checkActive()
        check(state == TransportSessionState.Connected) { "Transport session is not connected" }
        legacy.send(payload)
    }

    override suspend fun close() = gate.withLock {
        if (state == TransportSessionState.Closed || state == TransportSessionState.Idle) {
            state = TransportSessionState.Closed
            return@withLock
        }
        state = TransportSessionState.Closing
        try {
            legacy.disconnect()
            state = TransportSessionState.Closed
        } catch (error: Throwable) {
            state = TransportSessionState.Failed
            throw error
        }
    }

    suspend fun dispose() {
        if (disposed) return
        close()
        legacy.onPacketReceived = null
        disposed = true
    }

    private fun checkActive() = check(!disposed) { "Transport session is disposed" }
}
